### PR TITLE
Make work id and author id fields accept more types of ids

### LINKED ISSUE #6356: Make work id and author id fields on book edit page smarter about what they accept
On the book editing page, the fields for setting the work for an edition and for setting the author of a work have autosuggest functionality — you can start typing a value to get a list of matching records. You can also type a work or author ID to get a specific record, but if you enter anything besides the bare ID, it doesn't work.

When doing librarian work, I often find myself wanting to enter a specific ID in these fields, especially if I've been making a lot of changes and autosuggest doesn't have the right values because SOLR hasn't caught up yet. This involves finding the item I want, clicking on it, carefully extracting the appropriate ID from the browser location bar, and pasting that into the editing field.

What I'd really love to be able to do is just right-click an author or work link somewhere, copy the URL, paste it into the editing field, and have the autosuggest be smart enough to figure out what I'm trying to do and suggest the right ID.

### Proposal & Constraints
I propose upgrading the work and author fields so that they can extract the appropriate ID from a complex string (like a URL), saving librarians a lot of fussy clicking and selecting IDs.

I don't see any downsides here. OLIDs are well designed so that they're easily distinguishable from other text. The likelihood of a false match is extremely small, and if it were to happen, the user still gets feedback about the value they're about to save, so they can make any necessary corrections.

### Stakeholders
@mheiman

### PR DESCRIPTION
**Pull request by mheiman**
Closes #6356

Updates the behavior of the author and work fields on the book edit page so that the autosuggest will extract author and work ids from strings (in particular URLs) containing other text.

### Technical
This solution replaces the functions is_author_olid() and is_work_olid() in openlibrary/utils/__init__.py (they are only called in addbooks.py) with two new functions, find_author_olid_in_string() and find_work_olid_in_string(), which will return the first matching ID of the appropriate type in the passed string.

openlibrary/plugins/upstream/addbook.py is updated to use these new functions in place of the old ones.

### Testing
Copy an author URL (e.g. https://openlibrary.org/authors/OL3745151A/Alexandre_Dumas_fils) or a work URL (e.g. https://openlibrary.org/works/OL15707699W?edition=ia%3Aladameauxcaml00duma) and paste them into the work author field or the edition work field of the book editing page. Autosuggest should offer the correct author or work. 

### Stakeholders
@mheiman

Attribution Disclaimer: By proposing this pull request, I affirm to have made a best-effort and exercised my discretion to make sure relevant sections of this code which substantially leverage code suggestions, code generation, or code snippets from sources (e.g. Stack Overflow, GitHub) have been annotated with basic attribution so reviewers & contributors may have confidence and access to the correct context to evaluate and use this code.

### CODE REVIEW

**Review by jimman2003 — COMMENTED** [review]
Generally good, but I wonder about the regexes... Should we match OL\d+A and OL\d+W respectively , then deal with the None result?

**Comment by mheiman** [comment]
@jimman2003 can you clarify what you mean? I’m not sure what you’re suggesting doing differently.

File: openlibrary/utils/__init__.py [hunk 1]
```diff
@@ -127,20 +127,24 @@ def dicthash(d):
         return d
 
 
-author_olid_re = re.compile(r'^OL\d+A$')
-
-
-def is_author_olid(s):
-    """Case sensitive check for strings like 'OL123A'."""
-    return bool(author_olid_re.match(s))
+author_olid_embedded_re = re.compile(r'.*?(OL\d+A).*', re.IGNORECASE)
```
Comment (jimman2003) [inline openlibrary/utils/__init__.py:130]:
```suggestion
author_olid_embedded_re = re.compile(r'OL\d+A', re.IGNORECASE)
```

File: openlibrary/utils/__init__.py [hunk 2]
```diff
@@ -127,20 +127,24 @@ def dicthash(d):
...
 
+def find_author_olid_in_string(s):
+    try:
+        found = re.search(author_olid_embedded_re, s).group(1)
```
Comment (jimman2003) [inline openlibrary/utils/__init__.py:134]:
```suggestion
found=re.search(author_olid_embedded_re, s)
return found.upper if found else ""
```

**Review by jimman2003 — CHANGES_REQUESTED** [review]
A example of what I mean by the above, similar changes to the work regex too
https://docs.python.org/3/library/re.html#re.search

**Comment by mheiman** [comment]
Thanks, @jimman2003 — that does make things a little more concise. Python isn't my native language, so I'm still getting a feel for it. :)

re.search doesn't return a string, though, so you have to use found.group to get a string result.  I've tested and updated the PR based on your suggestions.

**Comment by jimman2003** [comment]
Yeah, forgot about the match object :) , thank you

File: openlibrary/utils/__init__.py [hunk 3]
```diff
@@ -127,20 +127,18 @@ def dicthash(d):
         return d
 
 
-author_olid_re = re.compile(r'^OL\d+A$')
+author_olid_embedded_re = re.compile(r'OL\d+A', re.IGNORECASE)
 
+def find_author_olid_in_string(s):
+    found = re.search(author_olid_embedded_re, s)
```
Comment (cclauss) [inline openlibrary/utils/__init__.py:133]: A nice spot for three [doctests](https://docs.python.org/3/library/doctest.html) so we have proof that the function works as we expected.  This also makes sure that no one breaks the function in future code changes.
```suggestion
    """
    >>> find_author_olid_in_string("OL28641A")
    '28641'
    >>> find_author_olid_in_string("oL2864a")
    '28641'
    >>> find_author_olid_in_string("OL7500661M")
    ''
    """
    found = re.search(author_olid_embedded_re, s)
```

**Review by cdrini — APPROVED** [review]
Agree with @cclauss ; some doctests would be nice here :) Putting on testing now. Otherwise code lgtm  --  confirmed the removed functions weren't used anywhere else.

File: openlibrary/utils/__init__.py (hunk 3 above)
Comment (cdrini) [reply openlibrary/utils/__init__.py:133]: plus 1! Although note the return should be `OL28641A`.

**Comment by cdrini** [comment]
Testing lgtm! Noticed it was showing the "Create new author record" option for eg URLs; @mheiman and fixed on a call. Pushing new code to testing now.

**Pull request merged**